ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,496 @@
+
+```ruby
+module Temporalio
+  class Client
+    # This can be splatted into the constructor if one option needs to be
+    # changed. This is a Struct and not Data because we support Ruby 3.1.
+    Options = Struct.new(...)
+
+    # Connect a client (i.e. Connection.new + Client.new)
+    def self.connect(target_host, namespace, runtime: nil, ...)
+      ...
+    end
+
+    # Create client with existing connection. This must be all keyword arguments
+    # and must match the Options struct.
+    def initialize(connection:, namespace:, ...)
+      ...
+    end
+
+    # Duped options 
+    def dup_options
+      ...
+    end
+
+    # Several properties such as connection, namespace, workflow_service, etc
+
+    # Returns workflow handle
+    def start_workflow(
+      workflow,
+      *args,
+      id:,
+      task_queue:,
+      ...
+    )
+      ...
+    end
+
+    # Returns result
+    def execute_workflow(
+      workflow,
+      *args,
+      id:,
+      task_queue:,
+      ...
+    )
+      ...
+    end
+
+    # Returns workflow handle
+    def workflow_handle(id, run_id: nil, ...)
+      ...
+    end
+
+    # Lots more
+  end
+
+  class Client
+    class Connection
+      Options = Struct.new(...)
+
+      # Perform connection
+      def initialize(target_host:, runtime: nil, ...)
+        ...
+      end
+
+      # Duped options 
+      def dup_options
+        ...
+      end
+
+      # Access to raw workflow service
+      def workflow_service
+        ...
+      end
+    end
+
+    class WorkflowHandle
+      # Wait for result
+      def result(follow_runs: true)
+        ...
+      end
+
+      # Wait for result
+      def result(follow_runs: true)
+        ...
+      end
+
+      # Lots more
+    end
+
+    class CloudOperationsClient
+      # Similar to client but for cloud ops
+    end
+  end
+end
+```
+
+* There are many things missing here, but the idea is clear.
+* We expect most users and most samples to just use `Client.connect`.
+  * 💭 Why not make users create a connection themselves instead of maintaining similar options sets across
+    `Connection.initialize` and `Client.connect`? We have found most people only need a single client per connection.
+* Users will have access to raw clients via `client.workflow_service` and similar
+* Workflow handles exist with all relevant calls.
+* Note that we use the kwargs constructor + options struct pattern so that a user can access duped options, change one,
+  and reconstruct via `Client.new(**my_new_options.to_h)`
+  * ❓Is there a better way to allow people to easily change an option? We didn't want a `with` pattern because, at
+    least for `Connection`, it requires a reconnect.
+* Must make clear in docs which things issue an RPC call (e.g. start workflow) vs ones that don't (e.g. get workflow
+  handle).
+
+### Workflows in Client
+
+Early thoughts on workflows are that they'll be Ractors. For the purposes of this phase, all that may matter is how
```

Comment:
I left a comment about this in the Google form as well, but I'm not sure if Ractors would be the right move here. They're still very experimental and still have some fundamental limitations:

- https://rubykaigi.org/2023/presentations/ko1.html
  > Ractor was introduced in Ruby 3.0 as an experimental feature to enable parallel programming in Ruby. Now Ractor is not widely used for many reasons. For example, from the perspective of a Ractor implementer, Ractor does not have enough performance and quality. In this talk, we will explain which features are not enough, why they are difficult, and how to solve these difficulties.

- https://rubykaigi.org/2024/presentations/ko1.html

  > This talk presents recent updates to Ractor, which enables parallel and concurrent programming on Ruby.
  >
  > Ractor still lacks fundamental features. For example, we cannot use “require” method and “timeout” methods on non-main Ractors because of synchronization and implementation issues. We will discuss such problems and how to solve them. From a performance point of view, we have introduced the M:N thread scheduler in Ruby 3.3 and we will show the performance analysis with recent improvements.


They also still output scary warnings, even in Ruby 3.3:

```
irb(main):001> Ractor.new { puts RUBY_VERSION }
(irb):1: warning: Ractor is experimental, and the behavior may change in future versions of Ruby! Also there are many implementation issues.
3.3.4
=> #<Ractor:#2 (irb):1 terminated>
```

I definitely think that Ractors could be a good solution in a future world where Ractors are more polished, but IMO for now it would be safer to follow the [`coinbase/temporal-ruby`](https://github.com/coinbase/temporal-ruby) SDK and use Fibers as the basis for scaling workflow executions.

Reply:
One of the biggest challenges in SDK development is sandboxing for determinism. There are two main impediments to determinism: state isolation and restrict API use (e.g. disk IO). Ignoring the latter, if we can get the former it may be worth it. JS is lucky in that it can spin up a new V8 instance. But it is a nightmare for Go and Java and .NET and other devs to remember that they cannot share state. We expended a lot of effort making a Python sandbox just to get this state isolation so that globals don't pollute.

We will have to evaluate performance to see if this can work, and we can provide a way to opt-out of Ractors (for the most part none of the Ractor stuff would be visible to users unless they wanted to do bad things like expose something shareable).

> Fibers as the basis for scaling workflow executions.

We will definitely still be using fibers (we will have our own fiber scheduler since we need deterministic coroutines), but if we can get state isolation by default, it may be worth the performance penalty. This is more of a phase 2 thing though, workflows aren't going to be done now.

Reply:
Got it, that makes sense! If it's only for workflow-internal things, I think Ractors sound great.

But if the goal is to isolate global state even between activities, I could see that being very challenging with Ractors. For example, I'm not sure if ActiveRecord would would play nicely with Ractors, because (AFAIK) the connection pool wouldn't be able to be shared between Ractors.

> Interestingly enough, deterministic isolated Temporal workflows is like one of the only good use cases I have seen for Ractors, heh.

Same, TBH 😅 

I _want_ to use them, but haven't really found a use case yet, especially given their current limitations :(

Reply:
> But if the goal is to isolate global state even between activities, I could see that being very challenging with Ractors

Definitely not. Activities will not have anything to do with Ractors at all. Activities are just normal Ruby code where you can do what you want (we will allow different executors so people can choose whether to run fiber/async vs threaded).

Reply:
Sounds great! Thanks for helping me understand!